### ISSUE TITLE
Crash on startup because mon_geom == None

### ISSUE THREAD

**Issue by 0ion9** [mypaint/mypaint#424]
Possibly due to a GTK3 api change?
I have only a single monitor, FWIW.

The following log illustrates what happens. It's slightly modified (the MON_NUM line, which is produced by a line I added to gui/workspace.py so that I could be sure that MyPaint wasn't asking for a silly monitor number)

```
INFO: mypaint: Installation layout: conventional POSIX-like structure with prefix u'/usr/local'
INFO: mypaint: There is an old-style configuration area in u'/home/kau/.mypaint/'
INFO: mypaint: Its contents can be migrated to $XDG_CONFIG_HOME and $XDG_DATA_HOME if you wish.
INFO: mypaint: See the XDG Base Directory Specification for info.
INFO: lib.i18n: POSIX: LANG='en_AU.UTF-8'
INFO: lib.i18n: POSIX: LANGUAGE=None
/usr/local/share/mypaint/gui/application.py:277: Warning: The property GtkWidget:margin-left is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(mypaint_main_xml)
/usr/local/share/mypaint/gui/application.py:277: Warning: The property GtkWidget:margin-right is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(mypaint_main_xml)
/usr/local/share/mypaint/gui/application.py:277: Warning: The property GtkMisc:xalign is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(mypaint_main_xml)
('MON_NUM', 0)
Traceback (most recent call last):
  File "/usr/local/bin/mypaint", line 462, in <module>
    main.main(datapath, iconspath, old_confpath, version=version)
  File "/usr/local/share/mypaint/gui/main.py", line 224, in main
    run()
  File "/usr/local/share/mypaint/gui/main.py", line 201, in run
    fullscreen = options.fullscreen,
  File "/usr/local/share/mypaint/gui/application.py", line 286, in __init__
    workspace.build_from_layout(self.preferences["workspace.layout"])
  File "/usr/local/share/mypaint/gui/workspace.py", line 271, in build_from_layout
    set_initial_window_position(toplevel_win, toplevel_pos)
  File "/usr/local/share/mypaint/gui/workspace.py", line 2142, in set_initial_window_position
    on_this_mon = (final_x < (mon_geom.x + mon_geom.width) and
AttributeError: 'NoneType' object has no attribute 'x'
```

Looking at the [API-docs](https://developer.gnome.org/gdk3/stable/GdkScreen.html#gdk-screen-get-monitor-geometry), it's not altogether clear what to put this down to. I do note the tooltip on allow-none (annotation of 'dest' param) is `NULL is OK, both for passing and for returning.`, which suggests a None returnvalue is normally a possibility and MyPaint may be required to handle the None case (probably by falling back to the screen dimensions).

GTK3 version is 3.17.8.r112.gcf0e087 (git)
OS is Arch Linux x86_64 and WM is AwesomeWM (3.5.6.659.g66c4ff7 (git)), FWIW.

**Comment by achadwick** [comment]
> falling back to the screen dimensions

Seems to be the most reasonable approach.

Come to think about it, we can use gdk_screen_get_monitor_workarea() instead of gdk_screen_get_monitor_geometry() now.

**Comment by 0ion9** [comment]
I have attempted a quick rewrite that does that. However, it segfaults(!)
```
    if None not in (final_x, final_y):
        on_existing_mon = False
        for mon_num in xrange(screen.get_n_monitors()):
            mon_geom = screen.get_monitor_workarea(mon_num)
            if mon_geom:
                on_this_mon = (final_x < (mon_geom.x + mon_geom.width) and
                               final_y < (mon_geom.x + mon_geom.height) and
                               final_x >= mon_geom.x and
                               final_y >= mon_geom.y)
            else:
                on_this_mon = (final_x < screen_w and
                               final_y < screen_h and
                               final_x >= 0 and
                               final_y >= 0)
            if on_this_mon:
                on_existing_mon = True
                break
        if not on_existing_mon:
            logger.warning("Calculated window position is offscreen; "
                           "ignoring %r" % ((final_x, final_y), ))
            final_x = None
            final_y = None
```

**Comment by achadwick** [comment]
Can it be tied down to `screen.get_monitor_workarea()` segfaulting? That said:

```
Python 2.7.10 (default, Jul  1 2015, 10:54:53) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from gi.repository import Gdk
>>> screen = Gdk.Screen.get_default()
>>> screen.get_monitor_workarea(1)
<CairoRectangleInt at 0xe4ad40>
>>> screen.get_monitor_workarea(2)

(.:11401): Gdk-CRITICAL **: gdk_screen_get_monitor_workarea: assertion 'monitor_num < gdk_screen_get_n_monitors (screen)' failed
<CairoRectangleInt at 0xe4ab70>
>>> screen.get_monitor_workarea(0)
<CairoRectangleInt at 0xe4ab80>
>>> 
```

works fine here.

Does [`DEBUGGING.md`](https://github.com/mypaint/mypaint/blob/bae0db944dc1dd2afd7a5026d36cd395bd4f372e/DEBUGGING.md) provide any help on your system? IIRC a scary message about gdb no longer supporting Python < 3.x when I upgraded one of my boxes recently (relevant: #392)

**Comment by 0ion9** [comment]
I tried your example above on Python3, just accidentally. FWIW 
**no** invocation of screen.get_monitor_workarea() returned a value; monitor indices > 0 gaves the 'CRITICAL' report you show, and monitor indice=0 appeared to do nothing.

.. and, I just tried it on Python 2.7.10; no difference, it never returns anything other than None.

gobject-introspection-runtime version = 1.44.0
GLib version 2.45.7.34 (git 4025b5a)

**Comment by achadwick** [comment]
@0ion9 I noticed a few other calls to Gdk.Screen.get_monitor_geometry() in there, so I've added a fuller workaround. Thanks for the initial rewrite, which I adapted.

If master is still segfaulting after this, could you try and track it down using the instructions in DEBUGGING.md please, and report it separately? Thanks.

**Comment by 0ion9** [comment]
@achadwick It's not segfaulting now, it just:

```
WARNING: gui.workspace: gdk_screen_get_monitor_geometry() returned NULL: using screen size instead as a fallback.
Traceback (most recent call last):
  File "/usr/local/bin/mypaint", line 462, in <module>
    main.main(datapath, iconspath, old_confpath, version=version)
  File "/usr/local/share/mypaint/gui/main.py", line 224, in main
    run()
  File "/usr/local/share/mypaint/gui/main.py", line 201, in run
    fullscreen = options.fullscreen,
  File "/usr/local/share/mypaint/gui/application.py", line 286, in __init__
    workspace.build_from_layout(self.preferences["workspace.layout"])
  File "/usr/local/share/mypaint/gui/workspace.py", line 284, in build_from_layout
    set_initial_window_position(toplevel_win, toplevel_pos)
  File "/usr/local/share/mypaint/gui/workspace.py", line 2112, in set_initial_window_position
    targ_geom = _get_target_area_geometry(screen, targ_mon_num)
  File "/usr/local/share/mypaint/gui/workspace.py", line 2214, in _get_target_area_geometry
    geom = Gdk.Rectangle()
TypeError: boxed cannot be created directly; try using a constructor, see: help(Gdk.Rectangle)
```

(currently running gtk3 3.17.8.r112.gcf0e087)

**Comment by achadwick** [comment]
Oh, yuck. Let's handle that over in #437 - this one is a direct consequence of my dud fix in 5ce9ff7.
